core/src/nodes/handled_node_tasks.rs

```diff
@@ -267,39 +270,57 @@ impl<TInEvent, TOutEvent, TIntoHandler, TReachErr, THandlerErr, TPeerId>
     }
 
     /// Provides an API similar to `Stream`, except that it cannot produce an error.
-    pub fn poll(&mut self) -> Async<HandledNodesEvent<TOutEvent, TIntoHandler, TReachErr, THandlerErr, TPeerId>> {
+    pub fn poll(&mut self) -> Async<HandledNodesEvent<TInEvent, TOutEvent, TIntoHandler, TReachErr, THandlerErr, TUserData, TPeerId>> {
+        let (message, task_id) = match self.poll_inner() {
+            Async::Ready(r) => r,
+            Async::NotReady => return Async::NotReady,
+        };
+
+        Async::Ready(match message {
+            InToExtMessage::NodeEvent(event) => {
+                HandledNodesEvent::NodeEvent {
+                    task: match self.tasks.entry(task_id) {
+                        Entry::Occupied(inner) => Task { inner },
+                        Entry::Vacant(_) => panic!("poll_inner only returns valid TaskIds; QED")
+                    },
+                    event
+                }
+            },
+            InToExtMessage::NodeReached(peer_id) => {
+                HandledNodesEvent::NodeReached {
+                    task: match self.tasks.entry(task_id) {
+                        Entry::Occupied(inner) => Task { inner },
+                        Entry::Vacant(_) => panic!("poll_inner only returns valid TaskIds; QED")
+                    },
+                    peer_id
+                }
+            },
+            InToExtMessage::TaskClosed(result, handler) => {
+                let (_, user_data) = self.tasks.remove(&task_id)
+                    .expect("poll_inner only returns valid TaskIds; QED");
+                HandledNodesEvent::TaskClosed {
+                    id: task_id, result, handler, user_data,
+                }
+            },
+        })
+    }
+
+    /// Since non-lexical lifetimes still don't work very well in Rust at the moment, we have to
+    /// split `poll()` in two. This method returns an `InToExtMessage` that is guaranteed to come
```

Comment:
I think you can still do without `poll_inner`:

```rust
    pub fn poll(&mut self) -> Async<HandledNodesEvent<TInEvent, TOutEvent, TIntoHandler, TReachErr, THandlerErr, TUserData, TPeerId>> {
        for to_spawn in self.to_spawn.drain() {
            tokio_executor::spawn(to_spawn);
        }
        loop {
            match self.events_rx.poll() {
                Ok(Async::Ready(Some((message, task_id)))) => {
                    // If the task id is no longer in `self.tasks`, that means that the user called
                    // `close()` on this task earlier. Therefore no new event should be generated
                    // for this task.
                    if !self.tasks.contains_key(&task_id) {
                        continue
                    }
                    return Async::Ready(match message {
                        InToExtMessage::NodeEvent(event) => {
                            HandledNodesEvent::NodeEvent {
                                task: match self.tasks.entry(task_id) {
                                    Entry::Occupied(inner) =>
                                        Task { inner },
                                    Entry::Vacant(_) =>
                                        panic!("contains_key returned true for this task ID; QED")
                                },
                                event
                            }
                        }
                        InToExtMessage::NodeReached(peer_id) => {
                            HandledNodesEvent::NodeReached {
                                task: match self.tasks.entry(task_id) {
                                    Entry::Occupied(inner) =>
                                        Task { inner },
                                    Entry::Vacant(_) =>
                                        panic!("contains_key returned true for this task ID; QED")
                                },
                                peer_id
                            }
                        }
                        InToExtMessage::TaskClosed(result, handler) => {
                            let (_, user_data) = self.tasks.remove(&task_id)
                                .expect("contains_key returned true for this task ID; QED");
                            HandledNodesEvent::TaskClosed {
                                id: task_id, result, handler, user_data
                            }
                        }
                    })
                }
                Ok(Async::NotReady) =>
                    return Async::NotReady,
                Ok(Async::Ready(None)) =>
                    unreachable!("The sender is in self as well, therefore the receiver never closes."),
                Err(()) =>
                    unreachable!("An unbounded receiver never errors")
            }
        }
    }
```

Reply:
Ideally I would have wanted to remove the `expect()` here.
Since we can't do that, I kind of prefer my version because of nicer code style.